hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/KeyDeletingService.java

```diff
@@ -234,32 +225,123 @@ private Pair<Integer, Boolean> submitPurgeKeysRequest(List<DeleteBlockGroupResul
         keyToUpdate.addAllKeyInfos(keyInfos);
         keysToUpdateList.add(keyToUpdate.build());
       }
+    }
+
+    int purgeKeyIndex = 0, updateIndex = 0, renameIndex = 0;
+    int currSize = 0;
+    boolean batchCapacityReached;
+
+    while (purgeKeyIndex < purgeKeys.size() ||
+        updateIndex < keysToUpdateList.size() ||
+        (renameEntriesToBeDeleted != null && renameIndex < renameEntriesToBeDeleted.size())) {
 
-      if (!keysToUpdateList.isEmpty()) {
-        purgeKeysRequest.addAllKeysToUpdate(keysToUpdateList);
+      batchCapacityReached = false;
+      PurgeKeysRequest.Builder requestBuilder = PurgeKeysRequest.newBuilder();
+
+      if (snapTableKey != null) {
+        requestBuilder.setSnapshotTableKey(snapTableKey);
       }
-    }
+      if (expectedPreviousSnapshotId != null) {
+        requestBuilder.setExpectedPreviousSnapshotID(
+            NullableUUID.newBuilder()
+                .setUuid(HddsUtils.toProtobuf(expectedPreviousSnapshotId))
+                .build());
+      }
+
+      currSize = requestBuilder.build().getSerializedSize();
+
+      // 3.1 Add purgeKeys
+      List<String> batchPurgeKeys = new ArrayList<>();
+
+      while (purgeKeyIndex < purgeKeys.size()) {
+        String nextKey = purgeKeys.get(purgeKeyIndex);
+
+        int estimatedKeySize = estimateStringEntrySize(nextKey);
 
-    OzoneManagerProtocolProtos.OMRequest omRequest = OzoneManagerProtocolProtos.OMRequest.newBuilder()
-        .setCmdType(OzoneManagerProtocolProtos.Type.PurgeKeys)
-        .setPurgeKeysRequest(purgeKeysRequest)
-        .setClientId(getClientId().toString())
-        .build();
-
-    // Submit PurgeKeys request to OM. Acquire bootstrap lock when processing deletes for snapshots.
-    try (BootstrapStateHandler.Lock lock = snapTableKey != null ? getBootstrapStateLock().lock() : null) {
-      OzoneManagerProtocolProtos.OMResponse omResponse = submitRequest(omRequest);
-      if (omResponse != null) {
-        purgeSuccess = purgeSuccess && omResponse.getSuccess();
+        if (currSize + estimatedKeySize > ratisLimit) {
+          batchCapacityReached = true;
+          break;
+        }
+
+        currSize += estimatedKeySize;
+        batchPurgeKeys.add(nextKey);
+        purgeKeyIndex++;
+      }
+
+      // Now actually add batchPurgeKeys
+      if (!batchPurgeKeys.isEmpty()) {
+        OzoneManagerProtocolProtos.DeletedKeys deletedKeys = OzoneManagerProtocolProtos.DeletedKeys.newBuilder()
+            .setVolumeName("").setBucketName("")
+            .addAllKeys(batchPurgeKeys).build();
+        requestBuilder.addDeletedKeys(deletedKeys);
+        currSize = requestBuilder.build().getSerializedSize();
+      }
+
+      // 3.2 Add keysToUpdate
+      while (!batchCapacityReached && updateIndex < keysToUpdateList.size()) {
+        OzoneManagerProtocolProtos.SnapshotMoveKeyInfos nextUpdate = keysToUpdateList.get(updateIndex);
```

Comment:
this itself is proto, why do we need build to temp and get size? we can directly get serialized size and compare.